### ISSUE TITLE
[PROPOSAL] Adopt c++17 as default standard in libs

### ISSUE THREAD

**Issue by Andreagit97** [falcosecurity/libs#792]
**Motivation**

Right now we don't have a well-defined cpp standard in our libraries. `c++17` could allow us to use new features like the [filesystem library](https://en.cppreference.com/w/cpp/filesystem) and it could allow adopters to rely on a recent standard during the libraries building phase. Moreover `c++17` supports quite old compilers so this shouldn't cause any issues. I've already tried a full `sinsp` and `Falco` build without any pitfalls.

Do you see any issues in doing this? @Molter73 @dwindsor @terylt @araujof @hbrueckner @gnosek @deepskyblue86

**Comment by Andreagit97** [comment]
if I forgot some adopters please feel free to tag them here :) ty

**Comment by Molter73** [comment]
From my side, I believe this wouldn't be a problem, we currently build the libraries with GCC 8 which supports C++17.

Think this would mostly affect people using GCC 4.8 on old distros like centos7/rhel7, (which are probably a lot more people than what we might imagine). IMO, we should try to support as many people as possible, but if the community feels that the newer features of C++17 will help improve the code, I would not oppose it.

As a follow up, would we want to create a roadmap for updating some outdated code to use C++17? Just so the impact of this decision isn't limited to newer code being added.

**Comment by dwindsor** [comment]
> Think this would mostly affect people using GCC 4.8 on old distros like centos7/rhel7

This is a big part of our use case.

**Comment by gnosek** [comment]
@dwindsor, there's a `devtoolset-9` series of packages, which we're successfully using to build the libs on CentOS 6. It seems it's also available for CentOS 7, so that might be an option for you

**Comment by deepskyblue86** [comment]
+1 on @Molter73's comment. Do we want to have a roadmap to migrate code to leverage C++17 goodies? If so, how do we want to track it? Issues with help-wanted label :slightly_smiling_face:?

**Comment by terylt** [comment]
We currently build with gcc8 and c++11.  I don't think it would be an issue to move to 17, but we could check.  For the filesystem stuff, we currently use a thirdparty lib that is basically a mirror of what is built into 17: https://github.com/gulrak/filesystem

**Comment by Andreagit97** [comment]
One of the next big steps will be to have public APIs in `sinsp` and having cpp17 available would allow us to have more effective weapons during the rewriting of some parts, for example, using `string_view` instead of our handmade implementation or using `filesystem` in situations like a  /proc scan. 

Obviously, we don't want to break anything, as @Molter73 correctly said we should try to support as many systems as possible, if we find a way to do it without too much pain, ok, otherwise we can wait until all our toolchains we will be updated. Thank you all for the feedback, and keep us updated if there are any changes (upgraded toolchains or something like that)

> +1 on @Molter73's comment. Do we want to have a roadmap to migrate code to leverage C++17 goodies? If so, how do we want to track it? Issues with help-wanted label slightly_smiling_face?

In the meanwhile this could be a great thing to do :+1:

**Comment by Andreagit97** [comment]
Wow, that's really good news :rocket:! Let's wait until the end of the Christmas holiday to see if anyone else has something to add :)

**Comment by federico-sysdig** [comment]
Christmas is over; I suggest we make sure all the stakeholders agree and, if so, proceed with the change. If you want I can help.
I remember `libs` code has no issue in compiling as C++17 or requires minimal corrections. The main change is in the gcc compiler switch that enforces the C++ standard. I'd even suggest to move away from that approach and use the variable `CMAKE_CXX_STANDARD`; it is less tied to compiler details.

**Comment by Andreagit97** [comment]
> Christmas is over; I suggest we make sure all the stakeholders agree and, if so, proceed with the change.

@federico-sysdig  no one added more comments so I think we can proceed!

> If you want I can help.

It would be amazing!

> I remember libs code has no issue in compiling as C++17 or requires minimal corrections. The main change is in the gcc compiler switch that enforces the C++ standard. I'd even suggest to move away from that approach and use the variable CMAKE_CXX_STANDARD; it is less tied to compiler details.

Completely agree with that, if you want to propose some changes to our cmake feel free to open a PR :)

**Comment by Andreagit97** [comment]
Moreover, I would add another point related to the c++ standard, our CMake version :upside_down_face: We all know that our CMake is quite ugly and confusing, moreover we don't have a clear well-defined minimum CMake version required!
Some examples:

1. In the root libs cmake we said minimum version is `2.8.5`
    
    ```
    cmake_minimum_required(VERSION 2.8.5)
    ```
    
    but this is not true! This is what happens if you build with Cmake `2.8.12.2` 
    
    ```
    [ 88%] No test step for 'googletest'
    [100%] Completed 'googletest'
    [100%] Built target googletest
    CMake Error at build/googletest-src/CMakeLists.txt:4 (cmake_minimum_required):
      CMake 3.5 or higher is required.  You are running version 2.8.12.2
    ```

2. building on Centos 7 with CMake `3.12.3` causes this! 
    
    https://github.com/falcosecurity/falco/issues/2293
    
    For this reason, we switched to CMake  `3.22.5`

TL; DR; I think we need to find a base valid version for our CMake initially I thought at something like `3.12.3` but this is not enough on Centos 7 for example, my dream would be something like `3.22` but maybe is too much :thinking: 

WDYT? @Molter73  @dwindsor @terylt @araujof @hbrueckner @gnosek @deepskyblue86 @Happy-Dude

**Comment by Andreagit97** [comment]
Any idea/suggestion around this cmake version?

**Comment by Molter73** [comment]
> Unfortunately, it seems that version of cmake in CentOS 7 is super old, so there's no working around that.

I just checked on a `centos:7` container with the `epel-release` repo enabled and the `cmake3` package is version `3.17.5`, so we should be good. Vanilla `cmake` package was never updated to version 3, so that is why it seems ancient 🙂

**Comment by deepskyblue86** [comment]
@Molter73 same on Amazon Linux 2, they keep separate cmake and cmake3 packages.

**Comment by Andreagit97** [comment]
Nope i think we can close this since we merged https://github.com/falcosecurity/libs/pull/950 :)
